app/src/main/java/com/github/polypoly/app/base/game/Game.kt

```diff
@@ -0,0 +1,160 @@
+        ) } }
+    val currentRound: Int = 1
+
+    /**
+     * Go to the next turn
+     */
+    fun nextTurn() {
+        // TODO update the data with the DB
+        if(isGameFinished()) {
+            val pastGame = endGame()
+            // TODO send the pastGame to the DB
+        }
+    }
+
+    /**
+     * Test if the game is finished
+     * @return true if the game is finished, false otherwise
+     * @throws IllegalStateException if the game mode is RICHEST_PLAYER and maxRound is null
+     */
+    private fun isGameFinished(): Boolean {
+        return when(rules.gameMode) {
+            GameMode.LAST_STANDING -> {
+                players.filter { !it.hasLose() }.size <= 1
+            }
+            GameMode.RICHEST_PLAYER -> {
+                if(rules.maxRound == null)
+                    throw IllegalStateException("maxRound can't be null in RICHEST_PLAYER game mode")
+                currentRound > rules.maxRound
+            }
+        }
+    }
+
+    /**
+     * Return the ranking of the players
+     * @return a [Map] of the [Player]s' [User] id and their rank
+     */
+    fun ranking(): Map<Long, Int> {
+        val playersSorted = players.sortedDescending()
+        val map = playersSorted.mapIndexed { index, player -> player.user.id to index+1 }.toMap().toMutableMap()
+        for(i in 1 until (players.size)) {
+            val currentPlayer = playersSorted[i]
+            val previousPlayer = playersSorted[i-1]
+            if(currentPlayer.compareTo(previousPlayer) == 0) {
+                map[previousPlayer.user.id]?.let{ map[currentPlayer.user.id] = it }
+            }
+        }
+        return map
+    }
+
+    /**
+     * End the game and return a PastGame object
+     * @return the PastGame object
+     * @throws IllegalStateException if the game is not finished
+     */
+    fun endGame(): PastGame {
+        if(!isGameFinished()) throw IllegalStateException("can't end the game now")
+        return PastGame(
+            users = players.map(Player::user),
+            usersRank = ranking().map { it.key to it.value }.toMap(),
+            date = dateBegin,
+            duration = System.currentTimeMillis() / 1000 - dateBegin,
+        )
+    }
+
+    /**
+     * Test if the user is playing in this game
+     * @param user the user to test
+     * @return true if the user is playing in this game, false otherwise
+     */
+    fun playInThisGame(user: User): Boolean {
+        return players.any { it.user.id == user.id }
+    }
+
+    /**
+     * Get the player associated to the user id
+     * @param userId the id of the user
+     * @return the player associated to the user id
+     */
+    fun getPlayer(userId: Long): Player? {
+        return players.find { it.user.id == userId }
+    }
+
+    /**
+     * Compute the winner of the bets, notify the players and update the balance of the players
+     * and the location owner.
+     */
+    fun computeAllWinnersOfBets() {
```

Comment:
Terminology-wise, I think you mean `bid`, which is the thing people do at auctions, rather than `bet`, which is a form of gambling

Reply:
Thank you for this terminology tip, I didn't know it! Fix done